File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();

        for (char bracket : s.toCharArray()) {
            if (bracket == '(' || bracket == '{' || bracket == '[') {
                stack.push(bracket);
            } else {
                if (stack.isEmpty()) {
                    return false;
                }

                char top = stack.pop();
                if ((bracket == ')' && top != '(') ||
                    (bracket == '}' && top != '{') ||
                    (bracket == ']' && top != '[')) {
                    return false;
                }
            }
        }

        return stack.isEmpty();
    }
}

File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        open_b = ['(', '{', '[']
        map_b = {
            ")": "(",
            "}": "{",
            "]": "["
        }
        st = []

        if len(s) % 2 != 0:
            return False
        
        for c in s:
            if c in open_b:
                st.append(c)
            else:
                if not st:
                    return False

                if map_b[c] == st[-1]:
                    st.pop()
                else:
                    return False
        return not st